docs:Module 4 (Running a chain) B9lab content updates 3-1 (#827)
cosmos/sdk-tutorials: b9lab-content/4-running-a-chain/04-node-api-and-cli.md

# Running a Node, API, and CLI

## Simapp

Run a blockhain and discover how to interact with it. 

Start with an educational sample called [Simapp](https://github.com/cosmos/cosmos-sdk/blob/master/simapp/).

Firstm create a folder and clone the cosmos-sdk into that folder:

```bash
$ mkdir cosmos
$ cd cosmos
$ git clone https://github.com/cosmos/cosmos-sdk
```

> Then build the cosmos-sdk:

Reviewer: It's literal, so I think the narrative correctly describes what follows:

```
$ cd cosmos-sdk
$ make build
```

As you might have expected, you do not have any keys yet. Add a new key:

```bash
$ ./simd keys add b9lab

- name: b9lab
  type: local
  address: cosmos1nw793j9xvdzl2uc9ly8fas5tcfwfetercpdfqq
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"A6TrsRO/OH91fAEFLohw7RwFB832NRsRWhQvE2t8cfLK"}'
  mnemonic: ""

**Important** write this mnemonic phrase in a safe place.
It is the only way to recover your account if you ever forget your password.

ivory uniform actual spot floor vessel monster rose yellow noise smile odor veteran human reason miss stadium phrase assault puzzle sentence approve coral apology
```

You can see the mnemonic at the end of the output:
```bash
ivory uniform actual spot floor vessel monster rose yellow noise smile odor veteran human reason miss stadium phrase assault puzzle sentence approve coral apology
```

Tis sequence of words is a mnemonic you can use to recover your public and private keys. In a production setting, the mnenonic should be stored in a reliable and confidential fashion such as a key-management infrastructure. 

Confirm that the key is has been added:

```bash
$ ./simd keys list
```

or with:

```bash
$ ./simd keys show b9lab
```

> A cosmos-sdk blockchain relies on validators to produce blocks. Validators are normally added and removed through an on-chain governance process, but initially there is no validator to generate the block in which such a transaction might exist. Your chain needs an account for bootstrapping purposes. 

Reviewer 2: ```suggestion
A Cosmos SDK blockchain relies on validators to produce blocks. Validators are added and removed through an on-chain governance process. Initially, there is no validator to generate the block in which such a transaction might exist. Your chain needs an account for bootstrapping purposes. 
```

Add your key (also know as an account) to the genesis file:

```bash
$ ./simd add-genesis-account b9lab 100000000stake
```

> The suffix, stake appended to the amount and configured in the genesis file is the unit for the tokens in this chain. This command `100000000` to your account. Check it with:

Reviewer 2: ```suggestion
The `stake` suffix is appended to the amount, configured in the genesis file, is established as the unit for the tokens in this chain. This command adds `100000000` stake to your account. Check the `genesis.json` file with:
```

```bash
$ cat ~/.simapp/config/genesis.json | grep -A 2 -B 2 stake
```

Because your blockchain is not actually running yet, you need to include bootstrap transactions in the genesis file.

> Note: In this scenario, you will need the 2/3 threshold for validation, so you will stake 70000000stake of our 100000000stake in the b9lab account we just created. 

Reviewer 2: ```suggestion
Note: In this scenario, you must meet the 2/3 threshold for validation, so you must stake 70000000stake of your 100000000stake in the b9lab account you just created. 
```

Reviewer 2: see how tricky it gets with different audiences (we or you)
be consistent and always use "you" as the second person audience

```

Now, start your blockchain:

```bash
$ ./simd start

6:23PM INF starting ABCI with Tendermint
6:23PM INF Starting multiAppConn service impl=multiAppConn module=proxy
6:23PM INF Starting localClient service connection=query impl=localClient module=abci-client
6:23PM INF Starting localClient service connection=snapshot impl=localClient module=abci-client
6:23PM INF Starting localClient service connection=mempool impl=localClient module=abci-client
6:23PM INF Starting localClient service connection=consensus impl=localClient module=abci-client

```

> You will see blocks producing and validating. Open a new terminal and check the balances:

Reviewer 2: ```suggestion
In the terminal window where you ran the command, you see blocks producing and validating. Open a new terminal to check the balances:
```

```bash
$ ./simd query bank balances $(./simd keys show b9lab -a)
balances:
- amount: "30000000"
  denom: stake
pagination:
  next_key: null
  total: "0"
```

> Create another key for a student and send it some tokens:

Reviewer 2: ```suggestion
Create another account for a student and send that account some tokens:
```

Reviewer 2: oh, are we using "key" or "account" (I think account is more commonly used in all the docs and tutorials to date) let's always use the same term, we explained key and account were the same earlier in this document

```bash
$ ./simd keys add student

- name: student
  type: local
  address: cosmos1m95dh3uc2s7fkn4w6v3ueux3sya96dhdudwa24
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"AgDYHucSs5vZ4viGyyoC0Qz6M7/+fEdqgOesEmeTdPE/"}'
  mnemonic: ""

**Important** record this mnemonic phrase in a safe place.
It is the only way to recover your account if you ever forget your password.

gown all scissors page panel table hill acoustic junior run winter cement mass clump moon adjust glare never satoshi easily illness hip rib multiply
```

Confirm the balance of the new account:

```bash
$ ./simd query bank balances $(./simd keys show student -a)
balances: []
pagination:
  next_key: null
  total: "0"
```

> This account does not have a balance. In fact, it is not in the blockchain yet. We need to send a transaction to change this new accounts balance:

Reviewer 2: ```suggestion
This account does not have a balance. In fact, the new account is not yet in the blockchain. You need to send a transaction to change this new account's balance:
```

```bash
$./simd tx bank send $(./simd keys show b9lab -a) $(./simd keys show student -a) 10stake --chain-id test-chain-rT4wZY

{"body":{"messages":[{"@type":"/cosmos.bank.v1beta1.MsgSend","from_address":"cosmos1nw793j9xvdzl2uc9ly8fas5tcfwfetercpdfqq","to_address":"cosmos1m95dh3uc2s7fkn4w6v3ueux3sya96dhdudwa24","amount":[{"denom":"stake","amount":"10"}]}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[],"gas_limit":"200000","payer":"","granter":""},"tip":null},"signatures":[]}

confirm transaction before signing and broadcasting [y/N]: y           
code: 0
codespace: ""
data: ""
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: ""
timestamp: ""
tx: null
txhash: D2CCFD91452F8C144BB1E7B54B9723EE3ED85925EE2C8AD843392721D072B895
```

> As you can see, it will ask you to confirm the signing and broadcasting and shows you usefull information such as gas_used.

Reviewer 2: ```suggestion
You are prompted to confirm the transaction before signing and broadcasting. The command output includes useful information such as `gas_used`.
```

> Now check the balance of the student again:

Reviewer 2: ```suggestion
Now check the balance of the student account again:
```

Reviewer 2: the student? hee hee 
(thinking of myself as a student, I had no balance!) just word choice here, let's be clear we're checking the account balance

```bash
./simd query bank balances $(./simd keys show student -a)
balances:
- amount: "10"
  denom: stake
pagination:
  next_key: null
  total: "0"
```

## CLI

> Inspect cosmos-sdk/simapp/simd/main.go:

Reviewer 2: ```suggestion
Inspect the `cosmos-sdk/simapp/simd/main.go` file:
```

```golang
package main

import (
  "os"

  "github.com/cosmos/cosmos-sdk/server"
  svrcmd "github.com/cosmos/cosmos-sdk/server/cmd"
  "github.com/cosmos/cosmos-sdk/simapp"
  "github.com/cosmos/cosmos-sdk/simapp/simd/cmd"
)

func main() {
  rootCmd, _ := cmd.NewRootCmd()

  if err := svrcmd.Execute(rootCmd, simapp.DefaultNodeHome); err != nil {
    switch e := err.(type) {
    case server.ErrorCode:
      os.Exit(e.Code)

    default:
      os.Exit(1)
    }
  }
}

```

> You can see, it imports github.com/cosmos/cosmos-sdk/simapp/simd/cmd for the CLI, which you can see at cosmos-sdk/simapp/simd/cmd/root.go:

Reviewer 2: ```suggestion
You can see imports for `github.com/cosmos/cosmos-sdk/simapp/simd/cmd` for the CLI that you can see in the `cosmos-sdk/simapp/simd/cmd/root.go` file:
```

# Running a Node, API, and CLI

## Simapp

Run a blockhain and discover how to interact with it. 

Start with an educational sample called [Simapp](https://github.com/cosmos/cosmos-sdk/blob/master/simapp/).

Firstm create a folder and clone the cosmos-sdk into that folder:

```bash
$ mkdir cosmos
$ cd cosmos
$ git clone https://github.com/cosmos/cosmos-sdk
```

Then build the cosmos-sdk:

```bash
$ cd cosmos-sdk
$ make build
```

This may take a few minutes. It will create a build folder and a simapp binary:

```bash

> $ ls build

Reviewer: It lists the contents of the `build` folder.

```

> Now, reset the database. The database will be in the initial state in any case if this is the first time you have used (which will be in the initial state anyway if this is the first time you are testing simapp:

Reviewer: Indeed. Fixed.

```bash
$ ./simd unsafe-reset-all

3:58PM INF Removed all blockchain history dir=/Users/b9lab/.simapp/data
3:58PM INF Generated private validator file keyFile=/Users/b9lab/.simapp/config/priv_validator_key.json stateFile=/Users/b9lab/.simapp/data/priv_validator_state.json
```

As you can see, it will list the files set to the initial state and their locations.

> Now, initialize the application:

Reviewer: Added a brief summary

```

In the output you can see that your chain_id is test-chain-rT4wZY. We could determine the chainid by passing the flag --chain-id.

Inspect the initial configuration:

```bash
$ cat ~/.simapp/config/genesis.json 
```

Inspect your keys:

```bash
$ ./simd keys list
[]

```

As you might have expected, you do not have any keys yet. Add a new key:

```bash
$ ./simd keys add b9lab

- name: b9lab
  type: local
  address: cosmos1nw793j9xvdzl2uc9ly8fas5tcfwfetercpdfqq
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"A6TrsRO/OH91fAEFLohw7RwFB832NRsRWhQvE2t8cfLK"}'
  mnemonic: ""

**Important** write this mnemonic phrase in a safe place.

> It is the only way to recover your account if you ever forget your password.

Reviewer 3: I don't understand the misunderstanding here.
The text shown here is the result of running `./simd keys add b9lab` more or less (when I tried it was in a different order, probably a version thing).

> It is the only way to recover your account if you ever forget your password.

Reviewer 4: @yaruwangway is this needed? As far as I'm aware you can just run `keys add` and the mnemonic is generated, no? https://docs.cosmos.network/master/run-node/keyring.html

ivory uniform actual spot floor vessel monster rose yellow noise smile odor veteran human reason miss stadium phrase assault puzzle sentence approve coral apology
```

You can see the mnemonic at the end of the output:
```bash
ivory uniform actual spot floor vessel monster rose yellow noise smile odor veteran human reason miss stadium phrase assault puzzle sentence approve coral apology
```

Tis sequence of words is a mnemonic you can use to recover your public and private keys. In a production setting, the mnenonic should be stored in a reliable and confidential fashion such as a key-management infrastructure. 

Confirm that the key is has been added:

```bash
$ ./simd keys list
```

or with:

```bash
$ ./simd keys show b9lab
```

A cosmos-sdk blockchain relies on validators to produce blocks. Validators are normally added and removed through an on-chain governance process, but initially there is no validator to generate the block in which such a transaction might exist. Your chain needs an account for bootstrapping purposes. 

> Add your key (also know as an account) to the genesis file:

Reviewer 3: Are Cosmos of the opinion that the words:
-  key describes the bits and bytes that are used to sign and identify.
- account describes the knowledge of such an identified key as saved in the application state.

If yes, I will adjust.

```bash
$ ./simd add-genesis-account b9lab 100000000stake
```

The suffix, stake appended to the amount and configured in the genesis file is the unit for the tokens in this chain. This command `100000000` to your account. Check it with:

```bash
$ cat ~/.simapp/config/genesis.json | grep -A 2 -B 2 stake
```

Because your blockchain is not actually running yet, you need to include bootstrap transactions in the genesis file.

Note: In this scenario, you will need the 2/3 threshold for validation, so you will stake 70000000stake of our 100000000stake in the b9lab account we just created. 

```bash
$ ./simd gentx b9lab 70000000stake --chain-id test-chain-rT4wZY
Genesis transaction written to "/Users/muratoener/.simapp/config/gentx/gentx-cf6bff39bb84da39d214138ebba8bcba4ccb848d.json"
```

> Update your genesis file:

Reviewer 3: Changed this line to explain. Please resolve if this is correct & enough.

```bash

> $ ./simd collect-gentxs

Reviewer 4: Added an extra line above. I think this should suffice?

6:23PM INF starting ABCI with Tendermint
6:23PM INF Starting multiAppConn service impl=multiAppConn module=proxy
6:23PM INF Starting localClient service connection=query impl=localClient module=abci-client
6:23PM INF Starting localClient service connection=snapshot impl=localClient module=abci-client
6:23PM INF Starting localClient service connection=mempool impl=localClient module=abci-client
6:23PM INF Starting localClient service connection=consensus impl=localClient module=abci-client

```

You will see blocks producing and validating. Open a new terminal and check the balances:

```bash
$ ./simd query bank balances $(./simd keys show b9lab -a)
balances:
- amount: "30000000"
  denom: stake
pagination:
  next_key: null
  total: "0"
```

Create another key for a student and send it some tokens:

```bash
$ ./simd keys add student

- name: student
  type: local
  address: cosmos1m95dh3uc2s7fkn4w6v3ueux3sya96dhdudwa24
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"AgDYHucSs5vZ4viGyyoC0Qz6M7/+fEdqgOesEmeTdPE/"}'
  mnemonic: ""

**Important** record this mnemonic phrase in a safe place.
It is the only way to recover your account if you ever forget your password.

> gown all scissors page panel table hill acoustic junior run winter cement mass clump moon adjust glare never satoshi easily illness hip rib multiply

Reviewer 3: Same, this is the result of the command line.

> gown all scissors page panel table hill acoustic junior run winter cement mass clump moon adjust glare never satoshi easily illness hip rib multiply

Reviewer 4: See convo above. It think this is fine based on https://docs.cosmos.network/master/run-node/keyring.html

```bash
$./simd tx bank send $(./simd keys show b9lab -a) $(./simd keys show student -a) 10stake --chain-id test-chain-rT4wZY

{"body":{"messages":[{"@type":"/cosmos.bank.v1beta1.MsgSend","from_address":"cosmos1nw793j9xvdzl2uc9ly8fas5tcfwfetercpdfqq","to_address":"cosmos1m95dh3uc2s7fkn4w6v3ueux3sya96dhdudwa24","amount":[{"denom":"stake","amount":"10"}]}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[],"gas_limit":"200000","payer":"","granter":""},"tip":null},"signatures":[]}

confirm transaction before signing and broadcasting [y/N]: y           
code: 0
codespace: ""
data: ""
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: ""
timestamp: ""
tx: null
txhash: D2CCFD91452F8C144BB1E7B54B9723EE3ED85925EE2C8AD843392721D072B895
```

As you can see, it will ask you to confirm the signing and broadcasting and shows you usefull information such as gas_used.

Now check the balance of the student again:

```bash
./simd query bank balances $(./simd keys show student -a)
balances:
- amount: "10"
  denom: stake
pagination:
  next_key: null
  total: "0"
```

## CLI

> Inspect cosmos-sdk/simapp/simd/main.go:

Reviewer 3: @anilder89 I made some changes, please tell me whether this is in keeping with what you wanted to express.

# Running a Node, API, and CLI

## Compile Simapp

Run a blockchain and discover how to interact with it.

> Start with an educational sample called [Simapp](https://github.com/cosmos/cosmos-sdk/blob/master/simapp/).

Reviewer 4: Just a little extra clarification on Simapp

```suggestion
The Cosmos SDK repository contains a folder called [Simapp](https://github.com/cosmos/cosmos-sdk/blob/master/simapp/). This contains the code to run a simulated version of the SDK so you can test commands without actually interacting with your chain. The binary is called `simd` and we'll be using that interact with our node for now.
```

        "inflation_max": "0.200000000000000000",
        "inflation_min": "0.070000000000000000",
        "inflation_rate_change": "0.130000000000000000",
        "mint_denom": "stake"
      }
    },
    "params": null,
    "slashing": {
      "missed_blocks": [],
      "params": {
        "downtime_jail_duration": "600s",
        "min_signed_per_window": "0.500000000000000000",
        "signed_blocks_window": "100",
        "slash_fraction_double_sign": "0.050000000000000000",
        "slash_fraction_downtime": "0.010000000000000000"
      },
      "signing_infos": []
    },
    "staking": {
      "delegations": [],
      "exported": false,
      "last_total_power": "0",
      "last_validator_powers": [],
      "params": {
        "bond_denom": "stake",
        "historical_entries": 10000,
        "max_entries": 7,
        "max_validators": 100,
        "unbonding_time": "1814400s"
      },
      "redelegations": [],
      "unbonding_delegations": [],
      "validators": []
    },
    "upgrade": {},
    "vesting": {}
  },
  "chain_id": "test-chain-rT4wZY",
  "gentxs_dir": "",
  "moniker": "demo",
  "node_id": "cf6bff39bb84da39d214138ebba8bcba4ccb848d"
}
```

</ExpansionPanel>

> In the output, you can find your `chain_id`, which here happens to be `test-chain-rT4wZY`. Make a note of the one you have as you will need it later on to determine the chain id, by passing it to simapp via the flag `--chain-id`.

Reviewer 4: ```suggestion
In the output, you can find your `chain_id`, which in our build happens to be called `test-chain-rT4wZY`. Make a note of the one you have as you will need it later on to determine the chain id, by passing it to simapp via the flag `--chain-id`.
```

You can inspect the initial configuration with:

```bash
$ cat ~/.simapp/config/genesis.json
```

## Prepare Your Account

You can also inspect your keys, as you will need them. These are held in the backend keyring, which, by default, is that of the operating system:

```bash
$ ./simd keys list
[]
```

> As you might have expected, you do not have any keys yet. Let's remedy that and add a new key:

Reviewer 4: ```suggestion
As you might have expected, you do not have any keys yet. Let's fix that and add a new key:
```

```bash
$ ./simd keys add b9lab

- name: b9lab
  type: local
  address: cosmos1nw793j9xvdzl2uc9ly8fas5tcfwfetercpdfqq
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"A6TrsRO/OH91fAEFLohw7RwFB832NRsRWhQvE2t8cfLK"}'
  mnemonic: ""

**Important** write this mnemonic phrase in a safe place.
It is the only way to recover your account if you ever forget your password.

ivory uniform actual spot floor vessel monster rose yellow noise smile odor veteran human reason miss stadium phrase assault puzzle sentence approve coral apology
```

> You can see the mnemonic at the end of the output:
> ```bash
> ivory uniform actual spot floor vessel monster rose yellow noise smile odor veteran human reason miss stadium phrase assault puzzle sentence approve coral apology
> ```

Reviewer 4: Don't think we need to add the mnemonic again?
```suggestion
You can see the mnemonic at the end of the output above.
```

This sequence of words is a mnemonic that you can use to recover your public and private keys. In a production setting, the mnemonic must be stored in a reliable and confidential fashion such as a key-management infrastructure.

Confirm that the key has been added:

```bash
$ ./simd keys list
```

or with:

```bash
$ ./simd keys show b9lab
```

## Make Yourself a Proper Validator

> As you know by now, a Cosmos SDK blockchain relies on identified validators to produce blocks. Validators are added and removed through an on-chain governance process, i.e. via transactions. Initially, there is no validator to generate any block in which such a transaction might be added. You are in a catch-22 situation. Your initialized, but unstarted, chain needs a genesis account and validator for bootstrapping purposes.

Reviewer 4: As far as I'm aware validators are not added or removed via a governance process, just automatically based on voting power. @xavierlepretre @rob-Hitchens 

```suggestion
As you know by now, a Cosmos SDK blockchain relies on identified validators to produce blocks. Initially, there is no validator to generate blocks. You are in a catch-22 situation. Your initialized and unstarted chain needs a genesis account and validator for bootstrapping purposes.
```

Make your key (also know as an account) have an initial balance in the genesis file:

```bash
$ ./simd add-genesis-account b9lab 100000000stake
```

Appended here to the amount is the `stake` suffix. This `stake` represents the unit for the tokens in this chain as per the genesis file. Therefore this command adds `100000000` `stake` to your account. If in doubt, you can confirm the proper suffix in the `genesis.json` file with:

```bash
$ grep -A 2 -B 2 denom ~/.simapp/config/genesis.json
```

You can also confirm in the genesis file itself that you have an initial balance:

```bash
grep -A 10 balances ~/.simapp/config/genesis.json
```

> With this initial balance, and before you actually run your blockchain, you still need to escape the catch-22 and include bootstrap transactions in the genesis file.

Reviewer 4: ```suggestion
With this initial balance, and before you actually run your blockchain, you still need to escape the catch-22 and include your bootstrap transactions in the genesis file.
```

> Note: In this scenario, you must meet the 2/3 threshold for validation, so you must stake `70000000stake` of your `100000000stake` in the `b9lab` account you just created. Don't forget to use your own `--chain-id`.

Reviewer 4: ```suggestion
Note: In this scenario, you must meet the 2/3 threshold for validation, so you must stake at least `70000000stake` of your `100000000stake` in the `b9lab` account you just created. But make sure to not use all your stake, so you still have tokens to pay for gas. Don't forget to use your own `--chain-id`.
```

```bash
$ ./simd gentx b9lab 70000000stake --chain-id test-chain-rT4wZY
Genesis transaction written to "/Users/muratoener/.simapp/config/gentx/gentx-cf6bff39bb84da39d214138ebba8bcba4ccb848d.json"
```

> After having created this genesis transaction into its own file, include it into your genesis file:

Reviewer 4: ```suggestion
After having created this genesis transaction into its own file, collect all the genesis transactions with `collect-gentxs` to include it into your genesis file:
```

```bash
$ ./simd query bank balances $(./simd keys show b9lab -a)
balances:
- amount: "30000000"
  denom: stake
pagination:
  next_key: null
  total: "0"
```

## Send a Transaction

Let's practice sending a transaction. For that, you are going to create another account named "student" and send it some tokens:

```bash
$ ./simd keys add student

- name: student
  type: local
  address: cosmos1m95dh3uc2s7fkn4w6v3ueux3sya96dhdudwa24
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"AgDYHucSs5vZ4viGyyoC0Qz6M7/+fEdqgOesEmeTdPE/"}'
  mnemonic: ""

**Important** record this mnemonic phrase in a safe place.
It is the only way to recover your account if you ever forget your password.

gown all scissors page panel table hill acoustic junior run winter cement mass clump moon adjust glare never satoshi easily illness hip rib multiply
```

Before sending it anything, confirm that the balance of the new account is absent:

```bash
$ ./simd query bank balances $(./simd keys show student -a)
balances: []
pagination:
  next_key: null
  total: "0"
```

> This account does not have a balance. In fact, the new account is not yet in the blockchain. You need to send a transaction to change this new account's balance:

Reviewer 4: ```suggestion
This account does not have a balance. In fact, the new account does not yet exist in your blockchain, only the key pair has been generated and stored in your keyring. You need to send a transaction to change this new account's balance:
```

```bash
$ ./simd tx bank send $(./simd keys show b9lab -a) $(./simd keys show student -a) 10stake --chain-id test-chain-rT4wZY

{"body":{"messages":[{"@type":"/cosmos.bank.v1beta1.MsgSend","from_address":"cosmos1nw793j9xvdzl2uc9ly8fas5tcfwfetercpdfqq","to_address":"cosmos1m95dh3uc2s7fkn4w6v3ueux3sya96dhdudwa24","amount":[{"denom":"stake","amount":"10"}]}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[],"gas_limit":"200000","payer":"","granter":""},"tip":null},"signatures":[]}

confirm transaction before signing and broadcasting [y/N]: y           
code: 0
codespace: ""
data: ""
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: ""
timestamp: ""
tx: null
txhash: D2CCFD91452F8C144BB1E7B54B9723EE3ED85925EE2C8AD843392721D072B895
```

You are prompted to confirm the transaction before signing and broadcasting. That's good. The command output includes useful information such as `gas_used`.

> Now check again the balance of the student account:

Reviewer 4: ```suggestion
Now, check the balance of the student account again:
```

```go
func NewRootCmd() (*cobra.Command, params.EncodingConfig)
```

In it, [basic properties](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/simapp/simd/cmd/root.go#L51-L53) such as the application name are defined:

```go
rootCmd := &cobra.Command{
    Use:   "simd",
    Short: "simulation app",
```

In addition, observe that cobra is imported and [used for the CLI](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/simapp/simd/cmd/root.go#L144-L155) to redirect:

```go
rootCmd.AddCommand(
    genutilcli.InitCmd(simapp.ModuleBasics, simapp.DefaultNodeHome),
    genutilcli.CollectGenTxsCmd(banktypes.GenesisBalancesIterator{}, simapp.DefaultNodeHome),
    genutilcli.MigrateGenesisCmd(),
    genutilcli.GenTxCmd(simapp.ModuleBasics, encodingConfig.TxConfig, banktypes.GenesisBalancesIterator{}, simapp.DefaultNodeHome),
    genutilcli.ValidateGenesisCmd(simapp.ModuleBasics),
    AddGenesisAccountCmd(simapp.DefaultNodeHome),
    tmcli.NewCompletionCmd(rootCmd, true),
    NewTestnetCmd(simapp.ModuleBasics, banktypes.GenesisBalancesIterator{}),
    debug.Cmd(),
    config.Cmd(),
)
```

> Also explain [`simapp/app.go`](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/simapp/app.go) in which each module and keykeeper will be imported. The first thing you will see is a considerable [list of modules](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/simapp/app.go#L33-L44) that are used by most cosmos-sdk apps:

Reviewer 4: The first sentence is a bit unclear. Do you mean to say, "Also, please look at ... " ?
```suggestion
Also, have a look at [`simapp/app.go`](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/simapp/app.go) in which each module and keykeeper will be imported. The first thing you will see is a considerable [list of modules](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/simapp/app.go#L33-L44) that are used by most cosmos-sdk apps:
```

```go
...
  "github.com/cosmos/cosmos-sdk/x/auth"
  "github.com/cosmos/cosmos-sdk/x/auth/ante"
  authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
  authsims "github.com/cosmos/cosmos-sdk/x/auth/simulation"
  authtx "github.com/cosmos/cosmos-sdk/x/auth/tx"
  authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
  "github.com/cosmos/cosmos-sdk/x/auth/vesting"
  "github.com/cosmos/cosmos-sdk/x/bank"
  bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
  banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
  "github.com/cosmos/cosmos-sdk/x/capability"
  capabilitykeeper "github.com/cosmos/cosmos-sdk/x/capability/keeper"
  capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
...
```

> The modules in the `/cosmos-sdk/x/` folder are maintained by the Cosmos team. To understand a module, the best way is to have a look at the according `spec` folder. Have a look at the [`cosmos-sdk/x/bank/spec/01_state.md`](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/x/bank/spec/01_state.md) to understand the state of the bank module which you used in this section.

Reviewer 4: Clarifying the decentralised workflow.
```suggestion
The modules in the `/cosmos-sdk/x/` folder are maintained by several organisations working on the Cosmos stack. To understand a module, the best way is to have a look at the respective `spec` folder. For example, have a look at the [`cosmos-sdk/x/bank/spec/01_state.md`](https://github.com/cosmos/cosmos-sdk/blob/d83a3bf92c9a84bddf3f5eb6692a1101c18b42f1/x/bank/spec/01_state.md) to understand the state of the bank module which you used in this section.
```

Also, would really appreciate a quick wrap-up of what was learned and a call to action under "Next Up". Not critical for beta-launch!
